### PR TITLE
Fix Export links for on-premise

### PR DESCRIPTION
**Pull request by tarzzz**
- [x] Test on on-prem

### CODE REVIEW

**Comment by tarzzz** [comment]
@n-riesco For your review.

File: app/components/Settings/Preview/Preview.react.js [hunk 1]
```diff
@@ -190,10 +194,27 @@ class Preview extends Component {
...
+            let link;
+
             if (!('error' in data)) {
-                plotlyLinks.unshift({type: type, url: data.url});
+                link = plotlyLinks.find((link) => link.type === type);
+                if (link) {
+                    // if exists, overwrite it:
+                    link.url = data.url;
+                }
+                else {
```
Comment (n-riesco) [inline app/components/Settings/Preview/Preview.react.js]: :cow:

File: backend/persistent/PlotlyAPI.js [hunk 2]
```diff
@@ -26,15 +26,34 @@ export function PlotlyAPIRequest(relativeUrl, {body, username, apiKey, accessTok
     });
 }
 
-export function newDatacache(payloadJSON, type) {
+export function newDatacache(payloadJSON, type, requestor) {
```
Comment (n-riesco) [inline backend/persistent/PlotlyAPI.js]: no need for `requestor` since this PR already adds it to `payloadJSON`

File: backend/routes.js [hunk 3]
```diff
@@ -589,10 +589,10 @@ export default class Servers {
         /* Plotly v2 API requests */
 
         server.post('/datacache', function getDatacacheHandler(req, res, next) {
-            const {payload, type: contentType} = req.params;
+            const {payload, requestor, type: contentType} = req.params;
```
Comment (n-riesco) [inline backend/routes.js]: No need for `requestor` since this PR already adds it to `payloadJSON`

**Comment by n-riesco** [comment]
@tarzzz could you give some context? why do we need https://github.com/plotly/falcon-sql-client/pull/313/commits/4661d7561e4cb9ea5142ad9b2a70c5a329aabbf0 and https://github.com/plotly/falcon-sql-client/pull/313/commits/a8ce08663fac19116a84b04189c8ff0cdc1ca614 ? why don't we use the api token from the headers?

File: backend/persistent/PlotlyAPI.js (hunk 2 above)
Comment (tarzzz) [reply backend/persistent/PlotlyAPI.js]: Thanks !! Will update it..

File: backend/routes.js (hunk 3 above)
Comment (tarzzz) [reply backend/routes.js]: Actually, Thats not True. The `payload` here just contains the CSV (or Plot) data.

See this: https://github.com/plotly/falcon-sql-client/pull/313/commits/4661d7561e4cb9ea5142ad9b2a70c5a329aabbf0#diff-70a485b1508d01fef5e75ca9aad77b28R182

`payload` and `requestor` are different Keys.

File: backend/persistent/PlotlyAPI.js (hunk 2 above)
Comment (tarzzz) [reply backend/persistent/PlotlyAPI.js]: Update: 

https://github.com/plotly/falcon-sql-client/pull/313#discussion_r156592084 is valid for here as well, so this `requestor` is needed.

File: backend/routes.js (hunk 3 above)
Comment (n-riesco) [reply backend/routes.js]: isn't the username also stored in `payload.requestor`?

Comment (tarzzz) [reply backend/routes.js]: It is not.

Comment (n-riesco) [reply backend/routes.js]: Oh! I see confused `payload` and `payloadJSON` :fearful:  .https://github.com/plotly/falcon-sql-client/blob/264658595575f84ba9dcea44043835b6e75ca46d/app/components/Settings/Preview/Preview.react.js#L182-L183

**Comment by n-riesco** [comment]
@chriddyp Could you see my questions on slack and review commits https://github.com/plotly/falcon-sql-client/commit/4661d7561e4cb9ea5142ad9b2a70c5a329aabbf0 and https://github.com/plotly/falcon-sql-client/commit/a8ce08663fac19116a84b04189c8ff0cdc1ca614 ?
Everything else looks OK to me.

File: backend/persistent/PlotlyAPI.js [hunk 4]
```diff
@@ -26,15 +26,34 @@ export function PlotlyAPIRequest(relativeUrl, {body, username, apiKey, accessTok
...
 
-    return fetch('https://plot.ly/datacache', {
+    const users = getSetting('USERS');
+    const user = users.find(
+        u => u.username === requestor
+    );
+    const apiKey = user.apiKey;
+    const accessToken = user.accessToken;
+
+    let authorization = '';
```
Comment (scjody) [inline backend/persistent/PlotlyAPI.js]: I'm OK with this for now, but please log an issue re: 🐫ing it (it copies the code in `PlotlyAPIRequest` pretty closely).

**Comment by scjody** [comment]
Since @chriddyp isn't around I've taken a look at the PR and https://github.com/plotly/falcon-sql-client/commit/4661d7561e4cb9ea5142ad9b2a70c5a329aabbf0 and https://github.com/plotly/falcon-sql-client/commit/a8ce08663fac19116a84b04189c8ff0cdc1ca614 in particular. Looks good enough for me, but an issue should be logged re: 🐫.

💃

**Comment by tarzzz** [comment]
Added an issue for cleanup with details here: https://github.com/plotly/streambed/issues/10460
Merging this..

**Pull request merged**